### PR TITLE
Provide error when no config was found

### PR DESCRIPTION
**Pull request by jonkerj**
**What this PR does / why we need it**:
The control loop currently segfaults when none of the config files was found. This commit makes it give an actual error, providing the user with useful feedback.

**Which issue(s) this PR fixes** 
If needed, I'll raise an issue, please let me know.

**Special notes for your reviewer** *(optional)*:

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -259,8 +260,8 @@ func GetFlatIPAM(isControlLoop bool, IPAM *types.IPAMConfig, extraConfigPaths ..
 			return flatipam, foundflatfile, err
 		}
 	}
-	var err error
-	return flatipam, foundflatfile, err
+
+	return flatipam, foundflatfile, errors.New("no config file was found")
```
Comment (maiqueb) [inline pkg/config/config.go:264]: thanks , good catch.

Would you mind also changing line [260](https://github.com/k8snetworkplumbingwg/whereabouts/pull/323/files#diff-a3d824da3c42420cd5cbb0a4a2c0e7b5bfddd819652788a0596d195dc6e31fa5R260) to `return flatipam, foundflatfile, nil` ? I think that's more explicit.

**Review by maiqueb — APPROVED** [review]

**Review by maiqueb — COMMENTED** [review]
Could you also add a couple of unit tests to https://github.com/k8snetworkplumbingwg/whereabouts/blob/master/pkg/config/config_test.go ?

Let me know if you need any help with it.

**Comment by jonkerj** [comment]
Sure, if I can bother you to look at #226 😉 (unrelated to this issue, but I think it deserves some attention too)

**Comment by jonkerj** [comment]
> Could you also add a couple of unit tests to https://github.com/k8snetworkplumbingwg/whereabouts/blob/master/pkg/config/config_test.go ?
> 
> Let me know if you need any help with it.

Well I tried to, but then I discovered that a lot (maybe all) tests in `config_test.go` are failing now, because `GetFlamIPAM` (which throws an error when no config files are present) is invoked from [`LoadIPAMConfig`](https://github.com/k8snetworkplumbingwg/whereabouts/blob/master/pkg/config/config.go#L56). I'm not really sure if `LoadIPAMConfig` needs a bit of refactoring to detect this type of error, or if the tests need rewriting, or maybe even something else.  

Do you have a hint for me?

**Comment by maiqueb** [comment]
> > Could you also add a couple of unit tests to https://github.com/k8snetworkplumbingwg/whereabouts/blob/master/pkg/config/config_test.go ?
> > Let me know if you need any help with it.
> 
> Well I tried to, but then I discovered that a lot (maybe all) tests in `config_test.go` are failing now, because `GetFlamIPAM` (which throws an error when no config files are present) is invoked from [`LoadIPAMConfig`](https://github.com/k8snetworkplumbingwg/whereabouts/blob/master/pkg/config/config.go#L56). I'm not really sure if `LoadIPAMConfig` needs a bit of refactoring to detect this type of error, or if the tests need rewriting, or maybe even something else.
> 
> Do you have a hint for me?

Yeah, so now that you return an error when there's no config present (I agree this should happen), the tests should be re-written to ensure this configuration is available to the tests.

Let's make sure @dougbtv is on board with the conceptual change first if you don't mind.

I can guide you on adapting the tests later on. 

Stay tuned.

**Comment by maiqueb** [comment]
FWIW, the changes you need to do in the tests are quite small.

Here's what you need to do:

```golang
       var tmpDir string // this must be accessible from the tests
 
       BeforeEach(func() {
               var err error
               tmpDir, err = os.MkdirTemp("", "whereabouts") // must create a tmp dir to store the configs
               Expect(err).NotTo(HaveOccurred())
       })

       AfterEach(func() {
               Expect(os.RemoveAll(tmpDir)).To(Succeed()) // must clean-up the generated configs
       })

      It("can load a basic config", func() {
		conf := `` // each test will have a different conf; you don't have to change anything here

		confPath := fmt.Sprintf("%s/%s", tmpDir, "whereabouts.conf") 
		Expect(os.WriteFile(confPath, []byte(conf), 0755)).To(Succeed()) // must persist the conf in the file-system

		ipamconfig, _, err := LoadIPAMConfig([]byte(conf), "", confPath) // must pass the new conf-path to the func; it only knows about the default paths
		Expect(err).NotTo(HaveOccurred())
		... // further expectations, each tests looks for different stuff
	})
```

**Review by dougbtv — APPROVED** [review]

**Comment by maiqueb** [comment]
@jonkerj alright! @dougbtv 's on board with this change.

Please update the failing tests in a similar way to what I'm proposing in [this link](https://github.com/k8snetworkplumbingwg/whereabouts/pull/323#issuecomment-1519884297).

Also please add a unit test that **expects** an error to be thrown - because the config on disk is **not** available.

**Comment by jonkerj** [comment]
Hi, sounds allright, I will look into that in a week or so.

**Comment by jonkerj** [comment]
I've updated my PR with a commit that fixes those tests. All is green on my side. I am recycling the filename `whereabouts.conf` throughout the tests, because (if I understand everything correct) it gets reinitialized every test (`BeforeEach` et al) or gets overwritten anyways.

File: pkg/config/config_test.go [hunk 2]
```diff
@@ -49,6 +67,11 @@ var _ = Describe("Allocation operations", func() {
 
 	})
 
+	It("throws an error when no flat-files are found", func() {
+		_, _, err := GetFlatIPAM(true, &types.IPAMConfig{}, "")
+		Expect(err).To(HaveOccurred())
```
Comment (maiqueb) [inline pkg/config/config_test.go:72]: please match against the specific error using the `MatchError` matcher.

File: pkg/config/config_test.go [hunk 3]
```diff
@@ -49,6 +67,11 @@ var _ = Describe("Allocation operations", func() {
 
 	})
 
+	It("throws an error when no flat-files are found", func() {
+		_, _, err := GetFlatIPAM(true, &types.IPAMConfig{}, "")
```
Comment (maiqueb) [inline pkg/config/config_test.go:71]: please omit the last attribute; [GetFlatIPAM](https://github.com/k8snetworkplumbingwg/whereabouts/blob/e12772b9ef977269cfabe8c079748ebeb550bb0e/pkg/config/config.go#L225) has a variadic signature :

```suggestion
		_, _, err := GetFlatIPAM(true, &types.IPAMConfig{})
```

**Review by maiqueb — CHANGES_REQUESTED** [review]
I only have two minor comments. Other than that, looks good.

File: pkg/config/config_test.go (hunk 2 above)
Comment (jonkerj) [reply pkg/config/config_test.go]: I've created a new error type for this and the tests matches on this specifically (preferred that over prefix/string matching)

File: pkg/config/config_test.go (hunk 3 above)
Comment (jonkerj) [reply pkg/config/config_test.go]: Sure, I missed that, oops

**Comment by maiqueb** [comment]
You have plenty of failed unit tests in `cmd/whereabouts_test.go` . 

You need to take care of them.

**Comment by jonkerj** [comment]
You are right, I need to fix the unit tests on way more places than I did. I only tested `github.com/k8snetworkplumbingwg/whereabouts/pkg/config`, previously, my bad.

**Comment by jonkerj** [comment]
OK, we are almost there. There are still 5 unit tests failing in `cmd/whereabouts_test.go`, and I'm again unsure how to proceed. In one of the tests (`allocate and release addresses on ADD/DEL`), `ipamConfig(..)` is called, which in turn calls `config.LoadIPAMConfig(..)`, which now throws an error.

I think simply ignoring `ConfigFileNotFoundError` in `cmd/whereabouts_tests.go` around lines 1451-1454 would mimic the previous behavior (as no config found would be ignored anyways), right?

Also: I feel a bit awkward about both passing `byte[](conf)` *and* write it to a file and pass that to `LoadIPAMConfig` (feels like it's not really DRY). How do you think about that?

**Comment by maiqueb** [comment]
> OK, we are almost there. There are still 5 unit tests failing in `cmd/whereabouts_test.go`, and I'm again unsure how to proceed. In one of the tests (`allocate and release addresses on ADD/DEL`), `ipamConfig(..)` is called, which in turn calls `config.LoadIPAMConfig(..)`, which now throws an error.
> 
> I think simply ignoring `ConfigFileNotFoundError` in `cmd/whereabouts_tests.go` around lines 1451-1454 would mimic the previous behavior (as no config found would be ignored anyways), right?

Yeah, that **was** the previous behavior. 

... but IMO, the goal of this PR is to get rid of the previous behavior and ensure an error is thrown when a config is not provided. This must be reflected in the unit tests.

IMO, you cannot just ignore the error; you should either do a proper setup of the tests,  **or** mock the *real* component, providing an implementation that does what you want. 

Maybe the later is simpler.

> 
> Also: I feel a bit awkward about both passing `byte[](conf)` _and_ write it to a file and pass that to `LoadIPAMConfig` (feels like it's not really DRY). How do you think about that?

The whole API could be improved to be honest ... 

If you choose to pursue that, please do it in a separate PR.

**Comment by jonkerj** [comment]
> OK, we are almost there. There are still 5 unit tests failing in `cmd/whereabouts_test.go`, and I'm again unsure how to proceed. In one of the tests (`allocate and release addresses on ADD/DEL`), `ipamConfig(..)` is called, which in turn calls `config.LoadIPAMConfig(..)`, which now throws an error.
> 
> I think simply ignoring `ConfigFileNotFoundError` in `cmd/whereabouts_tests.go` around lines 1451-1454 would mimic the previous behavior (as no config found would be ignored anyways), right?

I did something else: I `os.WriteFile`'d the mashalled config to a tempfile, similar to your earlier suggestion. As the function prototype of `ipamConfig` does not allow for passing errors currently (that needs significant rewrite of the 5 involved tests), it `nil`s out when anything fails (which cause the test to fail). Not the prettiest, but all tests pass now.

**Review by maiqueb — APPROVED** [review]
Thank you for this change, and your willingness to jump through all the hurdles I kept setting for you.

**Comment by jonkerj** [comment]
you're welcome!
